frontend/ui/widget/inputdialog.lua

```diff
@@ -561,7 +561,14 @@ function InputDialog:onShowKeyboard(ignore_first_hold_release)
     -- NOTE: There's no VirtualKeyboard widget instantiated at all when readonly,
     --       and our input widget handles that itself, so we don't need any guards here.
     --       (In which case, isKeyboardVisible will return `nil`, same as if we had a VK instantiated but *never* shown).
-    self._input_widget:onShowKeyboard(ignore_first_hold_release)
+    if Device:hasKeyboard() then
+        if G_reader_settings:nilOrTrue("virtual_keyboard_enabled") then
+            self._input_widget:onShowKeyboard(ignore_first_hold_release)
+        end
+        return
+    else
+        self._input_widget:onShowKeyboard(ignore_first_hold_release)
+    end
```

Comment:
You can probably simplify that to:
```lua
if Device:hasKeyboard() and G_reader_settings:isFalse("virtual_keyboard_enabled") then
    return
end
```

Reply:
that early return, would never run this 

```lua
self.keyboard_visible = self._input_widget:isKeyboardVisible()
```

not a problem?

Reply:
No idea, it's for you who test this to tell :)
May be you should also force `self.keyboard_visible = false` in InputDialog:init() `if Device:hasKeyboard() and G_reader_settings:isFalse("virtual_keyboard_enabled")`.
Or/and also may be set `self.keyboard_unused = Device:hasKeyboard() and G_reader_settings:isFalse("virtual_keyboard_enabled")` in :init() and use it here and there as needed.